ngmix/prepsfmom.py

```diff
@@ -0,0 +1,639 @@
+    # use the image which is in the msk
+    if drow != 0 or dcol != 0:
+        cen_phase = _compute_cen_phase_shift(drow, dcol, dim, msk=msk)
+        kim *= cen_phase
+
+    # build the flux, radial, plus and cross kernels / moments
+    # the inverse FFT in our convention has a factor of 1/n per dimension
+    # the sums below are inverse FFTs but only computing the values at the
+    # real-space center of the object (0 in our coordinate system).
+    # thus we code the factor of 1/n by hand
+    df = 1/dim
+    df2 = df * df
+    df4 = df2 * df2
+
+    # we only sum where the kernel is nonzero
+    fkf = kernels["fkf"]
+    fkr = kernels["fkr"]
+    fkp = kernels["fkp"]
+    fkc = kernels["fkc"]
+
+    mf = np.sum((kim * fkf).real) * df2
+    mr = np.sum((kim * fkr).real) * df2
+    mp = np.sum((kim * fkp).real) * df2
+    mc = np.sum((kim * fkc).real) * df2
+
+    # build a covariance matrix of the moments
+    # here we assume each Fourier mode is independent and sum the variances
+    # the variance in each mode is simply the total variance over the input image
+    # we need a factor of the padding to correct for something...
+    m_cov = np.zeros((4, 4))
+    tot_var *= eff_pad_factor**2
+    tot_var_df4 = tot_var * df4
+    kerns = [fkf / kpsf_im, fkr / kpsf_im, fkp / kpsf_im, fkc / kpsf_im]
+    conj_kerns = [np.conj(k) for k in kerns]
+    for i in range(4):
+        for j in range(i, 4):
+            m_cov[i, j] = np.sum((kerns[i] * conj_kerns[j]).real) * tot_var_df4
+            m_cov[j, i] = m_cov[i, j]
+
+    mom = np.array([mf, mr, mp, mc])
+
+    return mom, m_cov
+
+
+def _make_mom_res(mom, mom_cov):
+    # now finally build the outputs and their errors
+    res = {}
+    res["flags"] = 0
+    res["flagstr"] = ""
+    res["flux"] = mom[0]
+    res["mom"] = mom
+    res["mom_cov"] = mom_cov
+    res["flux_flags"] = 0
+    res["flux_flagstr"] = ""
+    res["T_flags"] = 0
+    res["T_flagstr"] = ""
+
+    # we fill these in later if T > 0 and flux cov is positive
+    res["flux_err"] = 9999.0
+    res["T"] = -9999.0
+    res["T_err"] = 9999.0
+    res["s2n"] = -9999.0
+    res["e1"] = 9999.0
+    res["e2"] = 9999.0
+    res["e"] = np.array([-9999.0, -9999.0])
+    res["e_err"] = np.array([9999.0, 9999.0])
+    res["e_cov"] = np.diag([9999.0, 9999.0])
+    res["mom_err"] = np.ones(4) * 9999.0
+
+    # handle flux-only
+    if np.diagonal(mom_cov)[0] > 0:
+        res["flux_err"] = np.sqrt(mom_cov[0, 0])
+        res["s2n"] = res["flux"] / res["flux_err"]
+    else:
+        res["flux_flags"] |= 0x40
+        res["flux_flagstr"] += 'zero or neg flux var;'
+
+    # handle flux+T only
+    if np.all(np.diagonal(mom_cov)[0:2] > 0):
+        if mom[0] > 0:
+            res["T"] = mom[1] / mom[0]
+            res["T_err"] = get_ratio_error(
+                mom[1], mom[0],
+                mom_cov[1, 1], mom_cov[0, 0], mom_cov[0, 1]
+            )
+        else:
+            # flux <= 0.0
+            res["T_flags"] |= 0x4
```

Comment:
Should we start naming these lags now?  We can also use the same ones for GaussMom and Admom

Reply:
Could be a different PR

Reply:
+1 to a different PR. I have kept the values consistent.

Reply:
https://github.com/esheldon/ngmix/issues/179